### PR TITLE
Review Cosmin

### PR DESCRIPTION
**Pull request by MichaelRotaru**

### CODE REVIEW

File: includes/languages/romana/modules/payment/twispay.php [hunk 1]
```diff
@@ -8,7 +8,9 @@
 /* Configuration */
 define('MODULE_PAYMENT_TWISPAY_TEXT_TITLE', 'Plateste in siguranta cu cardul | Twispay');
 define('MODULE_PAYMENT_TWISPAY_TEXT_PUBLIC_TITLE', 'Plateste in siguranta cu cardul | Twispay');
+define('MODULE_PAYMENT_TWISPAY_LOADING_TEXT', 'Se incarca...');
```
Comment (cosminbosutar) [inline includes/languages/romana/modules/payment/twispay.php]: The 'romanian' language folder is called 'romana' and not 'ro'?

File: includes/languages/english/modules/payment/twispay.php [hunk 2]
```diff
@@ -8,7 +8,9 @@
 /* Configuration */
 define('MODULE_PAYMENT_TWISPAY_TEXT_TITLE', 'Credit card secure payment | Twispay');
 define('MODULE_PAYMENT_TWISPAY_TEXT_PUBLIC_TITLE', 'Credit card secure payment | Twispay');
+define('MODULE_PAYMENT_TWISPAY_LOADING_TEXT', 'Loading...');
```
Comment (cosminbosutar) [inline includes/languages/english/modules/payment/twispay.php]: The 'english' language folder is called 'english' and not 'en'?

File: includes/languages/english/modules/payment/twispay.php [hunk 3]
```diff
@@ -59,6 +70,22 @@
...
 define('JSON_DECODE_ERROR_TEXT','Json decode error');
 define('NO_POST_TEXT','[RESPONSE-ERROR]: no_post');
+define('TABLE_HEADING_SUBSCRIPTION_STATUS_TEXT','Subscriptions status');
+define('BUTTON_CANCEL_SUBSCRIPTION_TEXT','Cancel Subscription');
+define('TEXT_ENABLE','Enable');
+define('TEXT_DISABLE','Disable');
+define('TEXT_DAY','daily');
+define('TEXT_WEEK','weekly');
+define('TEXT_MONTH','monthly');
+define('TEXT_YEAR','yearly');
```
Comment (cosminbosutar) [inline includes/languages/english/modules/payment/twispay.php]: Where are this used?

File: README.md [hunk 4]
```diff
@@ -30,10 +30,163 @@ TODO 1. Download the Twispay payment module from OScommerce Marketplace, where y
...
+Recurring status - Activate / Deactivate the recurring process.
+Recurring duration - how many times to repeat the payment. After the specified number of payments is reached the automatic cancel operation will be called
+Recurring cycles - The length of time period.
+Recurring frequency - The type of period
+Trial status - Activate / Deactivate the trial period.
+Trial cycles - The length of time period.
+Trial frequency - The type of period
+Trial price - the price for trial period
+
+!!! Note: The information about recurring product is not visible by customer. The administrator has the obligation to create a suggestive name like "Subscription Product" and a suggestive description that should contain the recurring information listed above.
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply README.md]: done.

File: README.md [hunk 5]
```diff
@@ -30,10 +30,163 @@ TODO 1. Download the Twispay payment module from OScommerce Marketplace, where y
...
+      <?php echo tep_draw_button(BUTTON_CANCEL_SUBSCRIPTION_TEXT, 'close', '#', 'primary'); ?>
+    <?php } ?>
+</td> ..."
+
+==== add after line 321 "... <td valign="top"><table border="0" width="100%" cellspacing="0" cellpadding="2">
+  <tr class="dataTableHeadingRow"> ...":
+
+"... <td class="dataTableHeadingContent"><?= TABLE_HEADING_SUBSCRIPTION_STATUS_TEXT; ?></td> ..."
+
+==== add before line 352 "... <td class="dataTableContent"><?php echo '<a href="' . tep_href_link(FILENAME_ORDERS, tep_get_all_get_params(array('oID', 'action')) . 'oID=' . $orders['orders_id'] . '&action=edit') . '">' . tep_image(DIR_WS_ICONS . 'preview.gif', ICON_PREVIEW) . '</a>&nbsp;' . $orders['customers_name']; ?></td>
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply README.md]: The code snippet from above is too common that's why i choose to refer to the line below.

File: ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php [hunk 6]
```diff
@@ -20,18 +20,18 @@ class Twispay_Status_Updater
...
-                                         , 'THREE_D_PENDING' => '3d-pending' /* Waiting for 3d authentication */
-                                         , 'EXPIRING' => 'expiring' /* The recurring order has expired */
-                                         ];
+                                       , 'IN_PROGRESS' => 'in-progress' /* Authorized */
+                                       , 'COMPLETE_OK' => 'complete-ok' /* Captured */
+                                       , 'COMPLETE_FAIL' => 'complete-failed' /* Not authorized */
+                                       , 'CANCEL_OK' => 'cancel-ok' /* Capture reversal */
+                                       , 'REFUND_OK' => 'refund-ok' /* Refund received */
+                                       , 'PARTIAL_REFUNDED' => 'partial-refunded' /* Partial refunded */
+                                       , 'TOTAL_REFUNDED' => 'total-refunded' /* Fully refunded */
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php]: Yes

File: ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php [hunk 7]
```diff
@@ -96,95 +121,137 @@ public static function updateStatus_backUrl($decrypted)
...
+                    Oscommerce_Order::updateStatus($orderId, MODULE_PAYMENT_TWISPAY_CHARGEDBACK_ORDER_STATUS_ID/*Chargedback*/, $allowSameStatusOverwrite, ORDER_CHARGEDBACK_NOTICE_TEXT.$comments);
+                    Twispay_Logger::log(LOG_OK_STATUS_CHARGEDBACK_TEXT . $orderId);
                 break;
 
                 case Twispay_Status_Updater::$RESULT_STATUSES['REFUND_OK']:
                     /* Add refund transaction to order status history. */
-                    Oscommerce_Order::updateStatus($order_id, -1/*Same status*/, ORDER_REFUNDED_NOTICE_TEXT.$decrypted['transactionId']);
-                    Twispay_Logger::log(LOG_OK_STATUS_REFUNDED_TEXT . $order_id);
-                    return ['success' => false];
+                    Oscommerce_Order::updateStatus($orderId, -1/*Same status*/, $allowSameStatusOverwrite, ORDER_REFUNDED_NOTICE_TEXT.$decrypted['transactionId']);
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php]: This case is executed only for refund confirmations that comes via IPN and is used just to log the transaction into order history. The status is already change after the CURL operation succeeds.

File: ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php [hunk 8]
```diff
@@ -96,95 +121,137 @@ public static function updateStatus_backUrl($decrypted)
...
+                    Twispay_Logger::log(LOG_OK_STATUS_REFUNDED_TEXT . $orderId);
                 break;
 
                 case Twispay_Status_Updater::$RESULT_STATUSES['PARTIAL_REFUNDED']:
                     /* Add partial refund to transaction history. */
-                    Oscommerce_Order::updateStatus($order_id, -1/*Same status*/, ORDER_REFUND_REQUESTED_NOTICE_TEXT.$decrypted['transactionId']);
-                    Twispay_Logger::log(LOG_OK_STATUS_REFUNDED_TEXT . $order_id);
-                    return ['success' => false];
+                    Oscommerce_Order::updateStatus($orderId, -1/*Same status*/, $allowSameStatusOverwrite, ORDER_REFUND_REQUESTED_NOTICE_TEXT.$decrypted['transactionId']);
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php]: The order status is not changing for partial refunds. It is just needed for order transaction to be logged.

File: ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php [hunk 9]
```diff
@@ -96,95 +121,137 @@ public static function updateStatus_backUrl($decrypted)
...
+                    Twispay_Logger::log(LOG_OK_STATUS_REFUNDED_TEXT . $orderId);
                 break;
 
                 case Twispay_Status_Updater::$RESULT_STATUSES['TOTAL_REFUNDED']:
                     /* Mark order as refunded. */
-                    Oscommerce_Order::updateStatus($order_id, MODULE_PAYMENT_TWISPAY_REFUNDED_ORDER_STATUS_ID/*Refunded*/, ORDER_REFUND_REQUESTED_NOTICE_TEXT.$decrypted['transactionId']);
-                    Twispay_Logger::log(LOG_OK_STATUS_REFUNDED_TEXT . $order_id);
-                    return ['success' => false];
+                    if ($orderRecurring) {
+                        Oscommerce_Order::updateStatus($orderId, -1/*Same status*/, $allowSameStatusOverwrite, ORDER_REFUND_REQUESTED_NOTICE_TEXT.$decrypted['transactionId']);
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php]: Order status is not changing if a transaction of a recurring order is refunded.

File: ext/modules/payment/twispay/js/twispay_user_actions.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+          'action': 'cancel'
+        },
+        success: function(data) {
+          /** if ajax call succeeded */
+          if (data['canceled'] == 1) {
+            alert(data['message']);
+            window.location.reload(true);
+          } else {
+            /** if ajax call failed */
+            alert(data['message']);
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply ext/modules/payment/twispay/js/twispay_user_actions.js]: it is reactivated after the "confirm" popup is closed or accepted

File: includes/languages/english/modules/payment/twispay.php (hunk 3 above)
Comment (MichaelRotaru) [reply includes/languages/english/modules/payment/twispay.php]: This is used for the select options of the fields on the product page.

File: ext/modules/payment/twispay/twispay_s2s.php [hunk 11]
```diff
@@ -82,10 +83,10 @@
...
 
     /** If transaction succeded */
-    if($status['success']){
-      die("OK");
-    }else{
-      die();
+    if ($status['success']) {
+        die("OK");
+    } else {
+        die();
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply ext/modules/payment/twispay/twispay_s2s.php]: is this ok? die("Failed");

File: admin/ext/modules/payment/twispay/js/twispay_actions.js [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+          'amount': refund_amount.toFixed(2),
+          'action': 'refund'
+        },
+        /** if ajax call succeeded */
+        success: function(data) {
+          if (data['refunded'] == 1) {
+            alert(data['message']);
+            window.location.reload(true);
+          } else {
+            /** if ajax call failed */
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply admin/ext/modules/payment/twispay/js/twispay_actions.js]: It is reactivated after confirmation popup.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (cosminbosutar) [reply README.md]: Alright

File: ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (cosminbosutar) [reply ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php]: Ok

File: ext/modules/payment/twispay/twispay_s2s.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (cosminbosutar) [reply ext/modules/payment/twispay/twispay_s2s.php]: die("Internal processing failure");

File: ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (cosminbosutar) [reply ext/modules/payment/twispay/helpers/Twispay_Status_Updater.php]: You are right

File: includes/languages/romana/modules/payment/twispay.php (hunk 1 above)
Comment (MichaelRotaru) [reply includes/languages/romana/modules/payment/twispay.php]: Yes

File: includes/languages/english/modules/payment/twispay.php (hunk 2 above)
Comment (MichaelRotaru) [reply includes/languages/english/modules/payment/twispay.php]: Yes

**Pull request merged**